File: forumClient/src/Dto/app/ThreadInfo.ts
import type { UserBasicInfo } from "../UserInfo"
import type { PostBasicInfo } from "./PostInfo"

export type Thread = {
    threadId: string,
    threadName: string,
    createdDate: Date,
    topicId: string,
    createdByUserId: string,
    numberOfPosts: number,
    hasNewPosts: boolean
}

export type ThreadBasicInfo = {
    thread: Thread
    totalPosts: number,
    newestMessage: PostBasicInfo,
    createdByUser: UserBasicInfo
}

export type ThreadFullInfo = {
    thread: Thread,
    totalPosts: number,
    posts: PostBasicInfo[],
    createdByUser: UserBasicInfo
}

export type CreateThreadRequest = {
    topicId: string,
    threadName: string,
    createdByUserId: string,
}

export type CreateThreadWithPostRequest = {
    topicId: string,
    threadName: string,
    createdByUserId: string,
    messageContent: string,
    
}

export type ThreadSummary = {
    totalThreads: number
}

File: server/RestApiServer/Dto/Forum/ThreadInfo.cs
using RestApiServer.Db;
using RestApiServer.Db.Users;
using RestApiServer.Dto.App;
using RestApiServer.Utils;

namespace RestApiServer.Dto.Forum
{
    public class ThreadBasicInfo
    {
        public required ThreadEntry Thread { get; set; } = null!;
        public required int TotalPosts { get; set; } = 0;
        public PostBasicInfo NewestMessage { get; set; } = null!;
        public UserBasicInfo CreatedByUser { get; set; } = null!;
    }

    public class ThreadFullInfo 
    {
        public required ThreadEntry Thread { get; set; } = null!;
        public required int TotalPosts { get; set; } = 0;
        public required List<PostBasicInfo> Posts { get; set; }
        //Represents the user that created the thread
        public required UserBasicInfo CreatedByUser { get; set; }
    }

    public class ThreadSummary
    {
        public required int TotalThreads { get; set; }
    }
}